docs/cpp_code_snippets/rapids_cmake_ctest_allocation.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <rapids_cmake_ctest_allocation.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <cstdlib>
+#include <numeric>
+#include <string>
+#include <string_view>
+
+namespace rapids_cmake {
+
+namespace {
+GPUAllocation noGPUAllocation() {
+  return GPUAllocation{-1, -1};
+}
+
+GPUAllocation parseCTestAllocation(std::string_view env_variable) {
+  std::string gpu_resources{std::getenv(env_variable.begin())};
+  // need to handle parseCTestAllocation variable being empty
+
+  // need to handle parseCTestAllocation variable not having some
+  // of the requested components
+
+  // The string looks like "id:<number>,slots:<number>"
+  auto id_start = gpu_resources.find("id:") + 3;
+  auto id_end = gpu_resources.find(",");
+  auto slot_start = gpu_resources.find("slots:") + 6;
+
+  auto id = gpu_resources.substr(id_start, id_end - id_start);
+  auto slots = gpu_resources.substr(slot_start);
+
+  return GPUAllocation{std::stoi(id), std::stoi(slots)};
+}
+
+std::vector<GPUAllocation> determineGPUAllocations() {
+  std::vector<GPUAllocation> allocations;
+  const auto *resource_count = std::getenv("CTEST_RESOURCE_GROUP_COUNT");
+  if (!resource_count) {
+    allocations.push_back(std::move(noGPUAllocation()));
+    return allocations;
+  }
+
+  const auto resource_max = std::stoi(resource_count);
+  for (int index = 0; index < resource_max; ++index) {
+    std::string group_env = "CTEST_RESOURCE_GROUP_" + std::to_string(index);
+    std::string resource_group{std::getenv(group_env.c_str())};
+    std::transform(resource_group.begin(), resource_group.end(),
+                   resource_group.begin(), ::toupper);
+
+    if (resource_group == "GPUS") {
+      auto resource_env = group_env + "_" + resource_group;
+      auto &&allocation = parseCTestAllocation(resource_env);
```

Comment:
Are we using the same clang-format settings as cudf? I’d like to adopt those if not. It can be a separate PR, and I’m happy to do that if you agree it is worth doing.

Reply:
`rapids-cmake` doesn't really have a clang-format setup since it generally has no c++ code. With this being in the documents I am fine running a formatter over it once and comitting it.

Reply:
I have updated with a default formatting. If we are interested in making sure all cpp/cuda code anywhere in rapids-cmake matches the RAPIDS / libcudf style we should do that in a followup PR ( I am not opposed to that )

Reply:
Latest round of changes also brings in a `.clang-format` file to the root of rapids-cmake, since now the GPU detection + json file writing is done by a C++ executable we build